### ISSUE TITLE
Introduce displayName placeholder for @ParameterizedTest name

### ISSUE THREAD

**Issue by ViToni** [junit-team/junit5#1496]
## Bug Report

Using:

```xml
<junit.jupiter.version>5.2.0</junit.jupiter.version>
<junit.platform.version>1.2.0</junit.platform.version>
```

Using `TestInfo.getDisplayName()` for regular tests yields the method name.
Using `TestInfo.getDisplayName()`  on `@ParameterizedTest` yields the injected parameter which is not very distinctive / descriptive by itself.

Some kind of FQN for tests would be more useful especially for test NOT executed in an IDE where one has a hierarchical view which gives this information more context.

Example:

```java
package tests;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInfo;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.EnumSource;

public class TestInfoTest {

    private static enum TestEnum {
        A, B, C;
    }

    private String testName = getClass().getSimpleName();

    @BeforeEach
    public void beforeEachTest(final TestInfo testInfo) {
        testName = testInfo.getDisplayName();
    }

    @Test
    public void foo() {
        logTestName();
    }

    @Test
    public void bar() {
        logTestName();
    }

    @ParameterizedTest
    @EnumSource(TestEnum.class)
    public void foo(final TestEnum testEnum) {
        logTestName();
    }

    @ParameterizedTest
    @EnumSource(TestEnum.class)
    public void bar(final TestEnum testEnum) {
        logTestName();
    }

    private void logTestName() {
        System.out.println(testName);
    }
}
```

Output:
```
bar()
[1] A
[2] B
[3] C
foo()
[1] A
[2] B
[3] C
```

Expected Output:

```
bar()
bar() - A
bar() - B
bar() - C
foo()
foo() - A
foo() - B
foo() - C
```

Even better output:

```
TestInfoTest.bar()
TestInfoTest.bar(A)
TestInfoTest.bar(B)
TestInfoTest.bar(C)
TestInfoTest.foo()
TestInfoTest.foo(A)
TestInfoTest.foo(B)
TestInfoTest.foo(C)
```

**Comment by sbrannen** [comment]
> Using `TestInfo.getDisplayName()`  on `@ParameterizedTest` yields the injected parameter which is not very distinctive / descriptive by itself.

That is intentional. A _display name_ is short by default and intended to be "displayed" in a UI.

The underlying `TestIdentifier` supplied to `TestExecutionListener` implementations provides access to both the `TestSource` and the "legacy reporting name". Thus, tools that generate reports have the ability to access the FQCN or FQMN if they desire.

And you as a developer have the option to override the default, generated display name for both `@Test` methods and `@ParameterizedTest` methods.

Note that `TestInfo` provides you access to the `java.lang.Class` and `java.lang.reflect.Method`, and you can do whatever you want with those.

I am therefore closing this issue as "works as designed".

**Comment by sbrannen** [comment]
Actually, now that I think about it in greater detail, I realize that although the default display name for parameterized tests indeed works by design, there is room for improvement here.

For example, a custom display name for a `@RepeatedTest` can optionally include the _display name_ of the repeated test method via the `{displayName}` placeholder.

Thus, if we were to introduce analogous support for `@ParameterizedTest(name=...)`, you could then annotate your parameterized test method as follows to achieve your desired "Expected Output".

```java
@ParameterizedTest(name = "{displayName} - {arguments}")
@EnumSource(TestEnum.class)
void foo(TestEnum testEnum) {
	logTestName();
}
```

In light of that, I am reopening this issue as an _enhancement_ proposal instead of a _bug_.

**Comment by sbrannen** [comment]
@junit-team/junit-lambda, thoughts?

**Comment by sbrannen** [comment]
Tentatively slated for 5.3 RC1 for _team discussion_.

**Comment by sbrannen** [comment]
FYI: I have rename this issue to address the redefined scope.

**Comment by sbrannen** [comment]
Note, however, that the method parameter types would also be included in the display name.

Thus, if we implement the above proposal for a `{displayName}` placeholder, the "Expected Output" we actually be the following

```
bar()
bar(TestEnum) - A
bar(TestEnum) - B
bar(TestEnum) - C
foo()
foo(TestEnum) - A
foo(TestEnum) - B
foo(TestEnum) - C
```

**Comment by ViToni** [comment]
Thanks for the swift response @sbrannen.

My personal approach is that it would be really helpful to have some useful defaults (being it an improved `TestInfo` or another class which can be injected easily) provided by JUnit itself.

The reasoning is that it's better to invest time and effort once into the API and provide something once which can be reused by many instead of putting many to do a tedious and boring task to change annotations or to force people create custom formatters which all might provide a similar result.

Would like to have my tests clean and lean out of the box :)

**Comment by sbrannen** [comment]
If you want to include the full class/method information, why don't you just use `testInfo.getTestMethod().get().toString() + " - " + testInfo.getDisplayName()` for whatever it is you're logging?

**Comment by sbrannen** [comment]
> Would like to have my tests clean and lean out of the box.

Perhaps you could describe what it is you are actually trying to achieve so that we can best consult you.

For example, it may well be that you would be better off implementing an extension, but without knowing your use case, it's difficult to provide you guidance.

**Comment by ViToni** [comment]
Sometimes I have classes with many test cases where I want to print results, or e.g. when I use a randomizer for fuzzy tests I want to print the seed so that I could reproduce the situation.
A one liner is often already enough information. What is missing is some easy way to get an identifier  of the test method called (especially a parametrized one).

Using the methods from `TestInfo` I came up with this (only for one parameter):
```
final StringBuilder sb = new StringBuilder();

testInfo.getTestClass().ifPresent(
        (clazz) -> {
            sb.append(clazz.getName());
            testInfo.getTestMethod().ifPresent(
                    (method) -> {
                        sb.append('.');
                        if (testInfo.getDisplayName().startsWith(method.getName())) {
                            sb.append(testInfo.getDisplayName());
                        } else {
                            sb.append(method.getName()).append('(');
                            sb.append(testInfo.getDisplayName().split(" ")[1]).append(')');
                        }
                    }
            );
        }
);

testName = sb.toString();
```

Output:
```
tests.TestInfoTest.bar()
tests.TestInfoTest.bar(A)
tests.TestInfoTest.bar(B)
tests.TestInfoTest.bar(C)
tests.TestInfoTest.foo()
tests.TestInfoTest.foo(A)
tests.TestInfoTest.foo(B)
tests.TestInfoTest.foo(C)
```
However it is not the most elegant solution and it works only for regular test methods and test methods with only one parameter (maybe not even there in corner cases).

Either I would have to  put it in some test utility class so I don't have to come up with it again and again or  it would have to be copied into each test class.

As the `TestInfo` already has the information (e.g. to create the display name) it would be great if it could provide some kind on convenience method which does the work for the user...

**Comment by sbrannen** [comment]
_in progress_

**Comment by sbrannen** [comment]
Note that it properly handles any number of arguments to a `@ParameterizedTest`.

**Comment by sbrannen** [comment]
Addressed in `master` in bf0369448c434db0704872823fd7b10312c6a409.
